csm_web/scheduler/views.py

```diff
@@ -1,267 +1,150 @@
-
-    @property
-    def list_permission_source(self):
-        # Only the Owner or Leader of this Profile should be able to view its attendance
-        return get_object_or_404(Profile, pk=self.kwargs["pk"])
+        queryset = Course.objects.filter(valid_until__gte=timezone.now().date())
+        if self.action == 'sections':
+            queryset = queryset.filter(enrollment_open__lte=timezone.now(), enrollment_closed__gt=timezone.now())
+        return queryset
 
+    @action(detail=True)
+    def sections(self, request, pk=None):
+        course = get_object_or_error(self.queryset, pk=pk)
+        return Response(SectionSerializer(course.section_set.all(), many=True).data)
 
-class SectionDetail(generics.RetrieveAPIView):
-    """
-    Responds to GET /sections/$ID with the corresponding section.
-    """
 
-    queryset = Section.objects.all()
+class SectionViewSet(*viewset_with('retrieve')):
     serializer_class = SectionSerializer
-    permission_classes = (permissions.IsAuthenticatedOrReadOnly, IsLeaderOrReadOnly)
 
     def get_object(self):
-        obj = get_object_or_404(self.get_queryset(), pk=self.kwargs["pk"])
-        self.check_object_permissions(self.request, obj)
-        return obj
+        return get_object_or_error(self.get_queryset(), **self.kwargs)
 
+    def get_queryset(self):
+        mentor_sections = Section.objects.filter(mentor__user=self.request.user)
+        student_sections = Section.objects.filter(students__user=self.request.user)
+        return (mentor_sections | student_sections).distinct()
+
+    @action(detail=True, methods=['get', 'put'])
+    def students(self, request, pk=None):
+        section = get_object_or_error(self.get_queryset(), pk=pk)
+        if request.method == 'GET':
+            return Response(StudentSerializer(section.students.filter(active=True), many=True).data)
+        # PUT
+        with transaction.atomic():
+            """
+            We reload the section object for atomicity. Even though we do not update
+            the section directly, any student trying to enroll must first acquire a lock on the
+            desired section. This allows us to assume that current_student_count is correct.
+            """
+            section = Section.objects.select_for_update().get(pk=section.pk)
+            if section.current_student_count >= section.capacity:
+                raise PermissionDenied("There is no space available in this section", status.HTTP_423_LOCKED)
+            try:  # Student dropped a section in this course and is now enrolling in a different one
+                student = Student.objects.get(active=False, section__course=section.course, user=request.user)
+                student.section = section
+                student.active = True
+                student.save()
+                return Response(status=status.HTTP_204_NO_CONTENT)
+            except Student.DoesNotExist:  # Student is enrolling in this course for the first time
```

Comment:
Wouldn't this case also cover a student attempting to change sections?

Reply:
I don't understand. Could you elaborate?

Reply:
If a student already in a section in this course attempts to enroll in a section, then the `Student.objects.get` query would error, right? So we'd end up in this branch, where we attempt to create a new student?

Reply:
No, the `Student.objects.get` query only errors if no matching object is found. No matching object found => the current user does not already have a student profile in this course. Which brings us to the except branch where we then create a student profile for them since they don't already have one.

Reply:
The `active=False` in the query should cause such an error afaik